File: src/url-structure.ts
import type { PageData, UrlStructureData, UrlStructureStats } from './types.js';

/**
 * Decompose a URL into structured components for analysis.
 */
export function analyzeUrlStructure(url: string): UrlStructureData {
  try {
    const parsed = new URL(url);
    const path = parsed.pathname;
    // Split path into segments, filtering out empty strings from leading/trailing slashes
    const rawSegments = path.split('/').filter(Boolean);
    const lastSegment = rawSegments.length > 0 ? rawSegments[rawSegments.length - 1] : '';

    // Extract file extension from last segment
    let fileExtension = '';
    if (lastSegment.includes('.')) {
      const ext = lastSegment.split('.').pop()!.toLowerCase();
      // Only treat common web extensions as file extensions
      if (['html', 'htm', 'php', 'asp', 'aspx', 'jsp', 'cgi', 'xml', 'json', 'pdf', 'js', 'css'].includes(ext)) {
        fileExtension = ext;
      }
    }

    // Extract query parameters (deduplicate keys, keep first value)
    const queryParams: Record<string, string> = {};
    parsed.searchParams.forEach((value, key) => {
      if (!(key in queryParams)) {
        queryParams[key] = value;
      }
    });
    const parameterCount = Object.keys(queryParams).length;

    return {
      scheme: parsed.protocol.replace(':', ''),
      hostname: parsed.hostname,
      port: parsed.port || (parsed.protocol === 'https:' ? '443' : '80'),
      pathDepth: rawSegments.length,
      pathSegments: rawSegments,
      lastSegment,
      queryParams,
      parameterCount,
      hasFragment: parsed.hash.length > 1,
      hasTrailingSlash: path.length > 1 && path.endsWith('/'),
      fileExtension,
    };
  } catch {
    return {
      scheme: '', hostname: '', port: '',
      pathDepth: 0, pathSegments: [], lastSegment: '',
      queryParams: {}, parameterCount: 0,
      hasFragment: false, hasTrailingSlash: false, fileExtension: '',
    };
  }
}

/**
 * Build aggregate URL structure statistics from all crawled pages.
 */
export function buildUrlStructureStats(pages: PageData[]): UrlStructureStats | null {
  const pagesWithData = pages.filter(p => p.urlStructure && p.urlStructure.hostname);
  if (pagesWithData.length === 0) return null;

  const depthDistribution: Record<number, number> = {};
  const directoryCount: Record<string, number> = {};
  const paramFrequency: Record<string, number> = {};
  const extensionCount: Record<string, number> = {};
  let totalDepth = 0;
  let maxDepth = 0;
  let trailingSlashCount = 0;
  let withParamsCount = 0;

  for (const page of pagesWithData) {
    const us = page.urlStructure!;

    // Depth distribution
    depthDistribution[us.pathDepth] = (depthDistribution[us.pathDepth] || 0) + 1;
    totalDepth += us.pathDepth;
    if (us.pathDepth > maxDepth) maxDepth = us.pathDepth;

    // Top-level directory distribution
    if (us.pathSegments.length > 0) {
      const dir1 = '/' + us.pathSegments[0];
      directoryCount[dir1] = (directoryCount[dir1] || 0) + 1;
    } else {
      directoryCount['/'] = (directoryCount['/'] || 0) + 1;
    }

    // Query parameter frequency
    for (const key of Object.keys(us.queryParams)) {
      paramFrequency[key] = (paramFrequency[key] || 0) + 1;
    }
    if (us.parameterCount > 0) withParamsCount++;

    // File extension
    const ext = us.fileExtension || '(none)';
    extensionCount[ext] = (extensionCount[ext] || 0) + 1;

    if (us.hasTrailingSlash) trailingSlashCount++;
  }

  const avgDepth = pagesWithData.length > 0 ? Math.round((totalDepth / pagesWithData.length) * 10) / 10 : 0;

  // Sort directories by count descending, take top 15
  const topDirectories = Object.entries(directoryCount)
    .sort(([, a], [, b]) => b - a)
    .slice(0, 15)
    .map(([dir, count]) => ({ directory: dir, count }));

  // Sort parameters by frequency descending, take top 15
  const topParameters = Object.entries(paramFrequency)
    .sort(([, a], [, b]) => b - a)
    .slice(0, 15)
    .map(([param, count]) => ({ parameter: param, count }));

  // Sort extensions by count descending
  const extensionDistribution = Object.entries(extensionCount)
    .sort(([, a], [, b]) => b - a)
    .map(([ext, count]) => ({ extension: ext, count }));

  return {
    totalUrls: pagesWithData.length,
    avgPathDepth: avgDepth,
    maxPathDepth: maxDepth,
    depthDistribution,
    topDirectories,
    topParameters,
    extensionDistribution,
    urlsWithParams: withParamsCount,
    urlsWithTrailingSlash: trailingSlashCount,
  };
}

File: internal/analysis/urlstructure.go
package analysis

import (
	"net/url"
	"sort"
	"strings"

	"github.com/micelio/micelio/internal/types"
)

var commonWebExtensions = map[string]bool{
	"html": true, "htm": true, "php": true, "asp": true, "aspx": true,
	"jsp": true, "cgi": true, "xml": true, "json": true, "pdf": true,
	"js": true, "css": true,
}

// AnalyzeURLStructure decomposes a URL into structured components.
func AnalyzeURLStructure(rawURL string) types.URLStructureData {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return types.URLStructureData{}
	}

	segments := strings.Split(strings.Trim(parsed.Path, "/"), "/")
	if len(segments) == 1 && segments[0] == "" {
		segments = nil
	}

	var lastSegment string
	if len(segments) > 0 {
		lastSegment = segments[len(segments)-1]
	}

	var fileExtension string
	if dotIdx := strings.LastIndex(lastSegment, "."); dotIdx >= 0 {
		ext := lastSegment[dotIdx+1:]
		if commonWebExtensions[strings.ToLower(ext)] {
			fileExtension = ext
		}
	}

	params := make(map[string]string)
	for k, v := range parsed.Query() {
		if _, exists := params[k]; !exists && len(v) > 0 {
			params[k] = v[0]
		}
	}

	return types.URLStructureData{
		Scheme:           parsed.Scheme,
		Hostname:         parsed.Hostname(),
		Port:             parsed.Port(),
		PathDepth:        len(segments),
		PathSegments:     segments,
		LastSegment:      lastSegment,
		QueryParams:      params,
		ParameterCount:   len(params),
		HasFragment:      parsed.Fragment != "",
		HasTrailingSlash: strings.HasSuffix(parsed.Path, "/") && len(parsed.Path) > 1,
		FileExtension:    fileExtension,
	}
}

// BuildURLStructureStats aggregates URL structure statistics across all pages.
func BuildURLStructureStats(pages []*types.PageData) *types.URLStructureStats {
	if len(pages) == 0 {
		return nil
	}

	depthDist := make(map[int]int)
	dirFreq := make(map[string]int)
	paramFreq := make(map[string]int)
	extFreq := make(map[string]int)
	totalDepth := 0
	maxDepth := 0
	urlsWithParams := 0
	urlsWithTrailingSlash := 0

	for _, p := range pages {
		data := AnalyzeURLStructure(p.URL)
		depthDist[data.PathDepth]++
		totalDepth += data.PathDepth
		if data.PathDepth > maxDepth {
			maxDepth = data.PathDepth
		}
		if data.ParameterCount > 0 {
			urlsWithParams++
		}
		if data.HasTrailingSlash {
			urlsWithTrailingSlash++
		}
		if data.FileExtension != "" {
			extFreq[data.FileExtension]++
		}
		if len(data.PathSegments) > 0 {
			dirFreq["/"+data.PathSegments[0]+"/"]++
		}
		for k := range data.QueryParams {
			paramFreq[k]++
		}
	}

	// Top 15 directories
	topDirs := sortedNamedCounts(dirFreq, 15)
	topParams := sortedNamedCounts(paramFreq, 15)
	extDist := sortedExtCounts(extFreq)

	avgDepth := 0.0
	if len(pages) > 0 {
		avgDepth = float64(totalDepth) / float64(len(pages))
	}

	return &types.URLStructureStats{
		TotalURLs:             len(pages),
		AvgPathDepth:          avgDepth,
		MaxPathDepth:          maxDepth,
		DepthDistribution:     depthDist,
		TopDirectories:        topDirs,
		TopParameters:         topParams,
		ExtensionDistribution: extDist,
		URLsWithParams:        urlsWithParams,
		URLsWithTrailingSlash: urlsWithTrailingSlash,
	}
}

func sortedNamedCounts(freq map[string]int, n int) []types.NamedCount {
	type kv struct {
		key   string
		count int
	}
	items := make([]kv, 0, len(freq))
	for k, v := range freq {
		items = append(items, kv{k, v})
	}
	sort.Slice(items, func(i, j int) bool {
		return items[i].count > items[j].count
	})
	if n > 0 && len(items) > n {
		items = items[:n]
	}
	result := make([]types.NamedCount, len(items))
	for i, item := range items {
		result[i] = types.NamedCount{Name: item.key, Count: item.count}
	}
	return result
}

func sortedExtCounts(freq map[string]int) []types.ExtensionCount {
	type kv struct {
		ext   string
		count int
	}
	items := make([]kv, 0, len(freq))
	for k, v := range freq {
		items = append(items, kv{k, v})
	}
	sort.Slice(items, func(i, j int) bool {
		return items[i].count > items[j].count
	})
	result := make([]types.ExtensionCount, len(items))
	for i, item := range items {
		result[i] = types.ExtensionCount{Extension: item.ext, Count: item.count}
	}
	return result
}
